presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/README.md

```diff
@@ -0,0 +1,148 @@
+to meets some expectation expressed in the `Matcher`.
+
+*A `Matcher` failing to match against a PlanNode it is applied to does not
+intrinsically mean that the plan does not match the expectations expressed by
+the test.* This is because the plan testing framework uses a pattern-matching
+like algorithm to allow tests to express their expectations flexibly with
+regards to changes in the structure of the tree by using the `anyTree()`
+pattern.  The `anyTree()` pattern looks for its child patterns in any of the
+children of the node it's applied to.
+
+Adding matchers to a `PlanMatchPattern` is usually done using one of the static
+methods in PlanMatchPattern.java.
+
+### Aliases
+
+At a high level, Presto plans represent data through `Symbols`. A
+`TableScanNode`, for instance, generates one symbol per column in the table as
+its output.  `ProjectNodes` can apply a calculation to some input `Symbol` and
+generate a new `Symbol` as output.
+
+If you are writing tests, you may care about what is happening to data in the
+actual plan. The simplest possible example of this would be asserting that a
+`SELECT name FROM people` query has an `OutputNode` with an output `Symbol`
+that actually represents the data in the `name` column of the `people` table.
+
+In order for the plan testing framework to be independent of how Presto
+allocates `Symbols`, `Symbols` are given an alias where they originate, and
+must be referenced by that alias further up the plan. The value being assigned
+to a `Symbol` can either be matched against something that is independent of a
+`Symbol` (e.g., a `ColumnHandle` in a `TableScanNode`) or a value that is
+dependent on a previously aliased symbol (e.g., an `Expression` in a
+`ProjectNode`). Matching against the right side of the assignment ensures that
+the plan matching framework is independent of the implementation of the
+`SymbolAllocator`.
+
+## Two types of tests
+
+### Testing planner code
+
+Planner code should be tested by comparing the unoptimized plan to a set of
+expectations. For various reasons related to how `Symbols` are handled in
+unoptimized plans, this is frequently impossible. The minimum set of optimizers
+that needs to be applied for the plan matching framework to function is likely
+to be some subset of the following:
+
+- `UnaliasSymbolReferences`
+- `PruneIdentityProjections`
+- `PruneUnreferencedOutputs`
+
+`TestQuantifiedComparison` illustrates unit testing planner code.
+
+### Testing optimizer code
+
+Optimizer code will generally need to be subjected to two types of tests:
+
+1. Unit tests on the minimum set of optimizers needed for the optimizer to
+   function.
+2. A test using the full set of optimizers that ensures that the optimizer
+   continues to function after the introduction of new optimizers or reordering
+   of existing optimizers.
+
+`TestMergeWindows` illustrates both types of tests and goes further into the
+rationale for both types of tests.
+
+## Internals
+
```

Comment:
Write a brief intro explaining what internals are (if necessary), and how they relate to...the framework? the tests?